rust/repositories.bzl

```diff
@@ -42,98 +78,333 @@ filegroup(
+        "toolchain_name_prefix": attr.string(),
+    },
+    implementation = _rust_toolchain_repository_impl,
+)
+
+"""Generates a toolchain-bearing repository that declares the toolchains from some other
+rust_toolchain_repository.
+
+Args:
+  name: A unique name for this rule
+  parent_workspace_name: The name of the other rust_toolchain_repository
+  exec_triple: The Rust-style target triple for the compilation platform
+  extra_target_triples: The Rust-style triples for extra compilation targets
+  toolchain_name_prefix: The per-target prefix expected for the rust_toolchain declarations in the
+                         parent workspace.
+"""
+
+rust_toolchain_repository_proxy = repository_rule(
+    attrs = {
+        "parent_workspace_name": attr.string(mandatory = True),
+        "exec_triple": attr.string(mandatory = True),
+        "extra_target_triples": attr.string_list(),
+        "toolchain_name_prefix": attr.string(),
+    },
+    implementation = _rust_toolchain_repository_proxy_impl,
+)
+
+def rust_repository_set(name, version, exec_triple, extra_target_triples, iso_date = None):
+    """Assembles a remote repository for the given toolchain params, produces a proxy repository
+    to contain the toolchain declaration, and registers the toolchains.
+
+    N.B. A "proxy workspace" is needed to allow for registering the toolchain (with constraints)
+    without actually downloading the toolchain.
+
+    Args:
+      name: The name of the generated repository
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+      exec_triple: The rust-style target that this compiler runs on
+      extra_target_triples: Additional rust-style targets that this set of toolchains
+                            should support.
+    """
+
+    rust_toolchain_repository(
+        name = name,
+        exec_triple = exec_triple,
+        extra_target_triples = extra_target_triples,
+        iso_date = iso_date,
+        toolchain_name_prefix = "toolchain_for",
+        version = version,
+    )
+
+    rust_toolchain_repository_proxy(
+        name = name + "_toolchains",
+        exec_triple = exec_triple,
+        extra_target_triples = extra_target_triples,
+        parent_workspace_name = name,
+        toolchain_name_prefix = "toolchain_for",
+    )
+
+    toolchain_name_template = "@{name}_toolchains//:toolchain_for_{triple}"
+    all_toolchain_names = [toolchain_name_template.format(
+        name = name,
+        triple = exec_triple,
+    )]
+    for extra_target_triple in extra_target_triples:
+        all_toolchain_names.append(toolchain_name_template.format(
+            name = name,
+            triple = exec_triple,
+        ))
+
+    # Register toolchains
+    native.register_toolchains(*all_toolchain_names)
 
 # Eventually with better toolchain hosting options we could load only one of these, not both.
```

Comment:
Not sure what this comment was referring to before, but looks rotted now for sure